tsl/src/nodes/vector_agg/function/functions.h

```diff
@@ -16,19 +16,19 @@ typedef struct
 	/* Size of the aggregate function state. */
 	size_t state_bytes;
 
-	/* Initialize the aggregate function state pointed to by agg_value and agg_isnull. */
-	void (*agg_init)(void *agg_state);
+	/* Initialize the aggregate function states. */
+	void (*agg_init)(void *restrict agg_states, int n);
```

Comment:
I see some potential confusion over the vector agg interface here:

- Some functions operate over the entire agg state array 
- While other functions operate over individual elements.
- It seems the semantics of some functions are changing (e.g., `agg_init` seems to have gone from single element to array. And, I think there's maybe a bug related to this).

I think we should pick one approach or the other, or make it very clear from the function when it is single element or the entire array. Right now we just pass around the agg_state pointer and it is unclear whether it should point to the start of the array or a particular element.

IMO, I think we should probably mostly make this interface work over the entire array. (It is a vector interface after all). Let's make that explicit.

To that end, I am also wondering if we can make this struct/class self-contained to alleviate some of the above issues?

For example, why not have the `agg_state` array and `n` part of the struct instead of separating them. Then all agg functions will always take the struct pointer as first argument and it will always have access to the start of the array and the size. No ambiguity. If a particular element is to be accessed or updated, the function takes an offset arg.

Reply:
The current concept is that aggregate function state is per-key. Then, there are functions for working with one state and many states in a contiguous array. The latter are not in this PR but in the hash aggregation one, except of the `init`. They have a suffix `many` and take different parameters. Functions for one state are required to generate more efficient code for full-batch aggregation, we could've done w/o them if this was less performance-sensitive.

Sounds like you want another thing that is "array of aggregate function states", that should combine an implementation of resizable array with an implementation of aggregate transition functions. Currently the "resizable array" part is done by a couple of lines in the hash grouping strategy, I think it's only going to complicate things if we somehow try to move these lines into a wrapper around every aggregate function.

Reply:
Alloca and resize functions don't need to be a wrapper around every agg function. I am just arguing for putting things together that I think belong together for clarity instead of spread out across different parts of the code. It would make the vectoragg state stuff more into a reusable library. It is already, more or less, but it can be improved further, IMO.

 What I mean is something like this:

```
typedef struct VectorAggState VectorAggState;

typedef struct
{
	/* Size of the aggregate function state. */
	size_t state_bytes;
    
	/* Initialize the aggregate function states. */
	void (*agg_init)(VectorAggState *state);

	/* Aggregate a given arrow array. */
	void (*agg_vector)(VectorAggState *state, const ArrowArray *vector, const uint64 *filter,
					   MemoryContext agg_extra_mctx);

	/* Aggregate a scalar value, like segmentby or column with default value. */
	void (*agg_scalar)(VectorAggState *state, Datum constvalue, bool constisnull, int n,
					   MemoryContext agg_extra_mctx);

	/* Emit a partial aggregation result. */
	void (*agg_emit)(VectorAggState *state, Datum *out_result, bool *out_isnull);
} VectorAggFunctions;


typedef struct VectorAggState
{
    const VectorAggFunctions *funcs;
    int num_states;
    unsigned char states[FLEXIBLE_ARRAY_MEMBER];
} VectorAggState;


/* Initialize the aggregate function states. A resize function would be analogous  */
VectorAggState *
vector_agg_state_create(const VectorAggFunctions *funcs, int num_states)
{
    VectorAggState *state = palloc(sizeof(VectorAggState) + funcs->state_bytes * num_states);
    state->funcs = funcs;
    state->num_states = num_states;
    funcs->agg_init(state);

    return state;
}

#define GET_AGGSTATE(state, i) \
    ((state)->states + (state)->funcs->state_bytes * i)

static void
minmax_init(VectorAggState *state)
{
    for (int i = 0; i < state->num_states; i++)
    {
        MinMaxState *mmstate = GET_AGGSTATE(state, i);
        ...
    }
}

const VectorAggFunctions minmax_funcs = {
    .state_bytes = sizeof(MinMaxState),
    .agg_init = minmax_init,
     ...    
};
````

The `VectorAggState` struct always keeps the state array, the size, length, and functions together instead of being separate. Passing this pointer to functions is easier and more clear than having these things passed separately. It also makes it much less likely to make a mistake w.r.t. indexing, size, etc.

If you want to update or work with a particular agg state that is possible to (pseuo-code):

```
static void
minmax_update(VectorAggState *state, int pos, int value)
{
        MinMaxState *mmstate = GET_AGGSTATE(state, pos);

        mmstate->min = value;
}
```

Reply:
Looks like in this interface, the number of states of different functions can differ? This is not correct. The states of one function are stored contiguously in memory, but their index for the same key must be the same.

Let's continue this discussion after we have the hash grouping policy, the requirements will be more clear.